Support different NFT contracts with `dao-voting-cw721-staked`

JakeHartnell opened the pull request:
Now works with `sg721` as well as some custom `cw721` contracts. :partying_face: 

Closes #716

JakeHartnell committed:
Support a wider variety of NFT contracts

Makes the interface for instantiating a new collection a bit more generic
to account for most types of cw721 and sg721 contracts.

JakeHartnell committed:
Improve test coverage

noahsaso commented on contracts/voting/dao-voting-cw721-staked/src/contract.rs:
```diff
@@ -110,28 +131,25 @@ pub fn instantiate(
             INITITIAL_NFTS.save(deps.storage, &initial_nfts)?;
 
             // Create instantiate submessage for NFT roles contract
```
i think this comment is wrong now

noahsaso commented on contracts/voting/dao-voting-cw721-staked/src/testing/tests.rs:
```diff
@@ -867,6 +916,40 @@ fn test_active_threshold_percentage_lte_0() {
+        "cw721_voting",
+        None,
+    )
+    .unwrap_err();
```
can you validate which specific error this returns

noahsaso reviewed (changes requested):
looks goooood to me, we should test it on both stargaze and juno testnets before merging

bekauz commented on contracts/voting/dao-voting-cw721-staked/src/contract.rs:
```diff
@@ -89,10 +89,31 @@ pub fn instantiate(
+            let sg721_instantiate_msg = from_binary::<sg721::InstantiateMsg>(&instantiate_msg);
+
+            // Update the minter to be this contract, reserialize into Binary
+            let instantiate_msg = match (cw721_instantiate_msg, sg721_instantiate_msg) {
```
feels like having some sort of structure here could be nice and would help keep the key contract methods short and concise. what do you think?
didn't test but i'm thinking something along the lines of:
```rust
enum NftInstantiateMsg {
    Cw721(cw721_base::InstantiateMsg),
    Sg721(sg721::InstantiateMsg),
    // any future 721 implementations?
}

impl NftInstantiateMsg {
    fn update_minter(&mut self, minter: &str) {
        match self {
            NftInstantiateMsg::Cw721(msg) => msg.minter = minter.to_string(),
            NftInstantiateMsg::Sg721(msg) => msg.minter = minter.to_string(),
        }
    }
}

fn try_deserialize(instantiate_msg: Binary) -> Result<NftInstantiateMsg, ContractError> {
    if let Ok(cw721_msg) = from_binary::<cw721_base::InstantiateMsg>(&instantiate_msg) {
        return Ok(NftInstantiateMsg::Cw721(cw721_msg));
    }

    if let Ok(sg721_msg) = from_binary::<sg721::InstantiateMsg>(&instantiate_msg) {
        return Ok(NftInstantiateMsg::Sg721(sg721_msg));
    }
    
    Err(ContractError::NftInstantiateError {})
}

```

JakeHartnell replied:
Oh, love this.

JakeHartnell committed:
Address PR comments, check errors

JakeHartnell committed:
Apply @bekauz's suggestion

JakeHartnell committed:
Add migrate message

New features don't require any state migrations. Both `ACTIVE_THRESHOLD` and `HOOKS` are uninitialized if none are set, so no migration logic needed.

noahsaso reviewed (approved):
hell yes this looks so good!!

JakeHartnell committed:
More test coverage
